python/ray/data/datasource/datasink.py

```diff
@@ -0,0 +1,77 @@
+from typing import Any, Iterable, List
+
+from ray.data._internal.execution.interfaces import TaskContext
+from ray.data.block import Block
+
+WriteResult = Any
+
+
+class Datasink:
+    def on_write_start(self, **_) -> None:
+        """Callback for when a write job starts.
+
+        Use this method to perform setup for write tasks. For example, creating a
+        staging bucket in S3.
+
+        Args:
+            _: Forward-compatibility placeholder.
+        """
+        pass
+
+    def write(
+        self,
+        blocks: Iterable[Block],
+        ctx: TaskContext,
```

Comment:
Hmm don't we need to pass a pathname / pathname provider?

Reply:
No. Paths are specific to file-related IO, so it's not exposed in the base interface. For example, the concept of a filename provider wouldn't apply to BigQuery.

`FilenameProvider` is exposed in `RowBasedFileDatasink` and `BlockBasedDatasink`: https://github.com/ray-project/ray/pull/40691

Reply:
Ah... I'm a bit confused, how do we know which file to write the blocks to?

Reply:
If you're working with files, you'd subclass the `RowBasedFileDatasink` or `BlockBasedFileDatasink` base classes. In this case, you'd receive the file as an argument to `write_block_to_file(block, file)`. This is similar to how we previous passed the file to `FileBasedDatasource._write_block`.